### PR TITLE
Introduce AllocationLimit

### PR DESCRIPTION
**Pull request by beyeler**
Following your input in https://github.com/marketdesignresearch/mechlib/issues/22 I implemented a proposal for AllocationLimits:

1. I agree AllocationLimits and ActivityRules are not the same. There are yet some ActivityRules implemented for MLCA with Bounds, but I agree that this should be done in mechlib in a general way
2. AllocationLimits are known ex-ante and therefore also affect the efficient Allocation. Therefore it makes sense for me to add the AllocationLimits to the Domain. With this it is easy to maintain consistency throughout the experiments. I.e. you create a Domain with the AllocationLimits you want all Auctions and WDP's will use these AllocationLimits by default. We may still provide a method to access the Domain without the AllocationLimit (like I did for the getValue() function. But in general I would prefer to have a consistent implementation.
3. I agree and implemented it in this way. However, applying the AllocationLimit in different context needs some special knowledge about the Context, i.e. sats WDP, mechlib WDP, RandomSampling etc. To allow an easy implementation there I suggest we allow only one AllocationLimit per Bidder. You still can combine them, by introducing a new combined AllocationLimit, like in the example BundleSizeAndGoodAllocationLimit. 

Note that you will use the appropriate branch in sats. Also I added some more details from original GSVM: 
"Activity and purchase limits were such thatregional bidders can acquire at most four licenses, and the national bidder can acquireup to twelve licenses on the larger circle."
Therefore I added an AllocationLimit for the national bidder, such that he is only allowed to purchase licenses on the larger national circle. Furthermore, AllocationLimits are not yet completly integrated into sats. I.e. the getValueFunction() of the GSVMBidder might needs to respect the AllocationLimit by default as well. (Note that this function is used to sampel bundles for the starting price generation of CCA)

### CODE REVIEW

File: src/main/java/org/marketdesignresearch/mechlib/core/bidder/Bidder.java [hunk 1]
```diff
@@ -45,11 +46,30 @@
...
+     * the allocation limit is respected and a bundle is queried that can not be allocated
+     * to this bidder an exception is thrown. This is done in order to verify that you are 
+     * sure you want to query a bundle that can not be allocated to this bidder. In most 
+     * cases you probably won't query such a bundle. 
+     * 
+     * @param bundle the bundle
+     * @param ignoreAllocationLimits specifies weather the allocation limit should be respected.
+     * @return the value of this bidder for the bundle of goods
+     */
+    BigDecimal getValue(Bundle bundle, boolean ignoreAllocationLimits);
```
Comment (islerfab) [inline src/main/java/org/marketdesignresearch/mechlib/core/bidder/Bidder.java:72]: There is the `getValue(Bundle bundle, Bundle alreadyWon)` method that needs to be taken into account as well.

File: src/main/java/org/marketdesignresearch/mechlib/winnerdetermination/allocationlimit/OrWDPAllocationLimitUtils.java [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+import edu.harvard.econcs.jopt.solver.mip.Constraint;
+import edu.harvard.econcs.jopt.solver.mip.Variable;
+
+public enum OrWDPAllocationLimitUtils {
+	PROCESSOR;
+	
+	@SuppressWarnings("rawtypes")
+	private Map<Class<? extends AllocationLimit>, OrWDPAllocationLimitProcessor> processor = new HashMap<>();
+	
+	private OrWDPAllocationLimitUtils() {
```
Comment (islerfab) [inline src/main/java/org/marketdesignresearch/mechlib/winnerdetermination/allocationlimit/OrWDPAllocationLimitUtils.java:23]: Same concern as above

**Comment by beyeler** [comment]
I'm sure you can implement this in a more generic way. Thus you would need to make some assumptions that AllocationLimits can all be expressed as Constraints on allocated good (variables). However, I'm not sure how efficient this is. E.g. in the BidBasedWinnerDetermination there are currently no variables for goods. Also the way I implemented the OrWDP does not encode goods directly. I'm quite sure that it is possible but I don't know if it is worth the effort at the moment and I'm not sure about the performance (i.e. also for CCG prices where many WDS's need to be solved.

About the code smells. Java Maps have some problems working with generics. Thus you somewhen need to cast the generic argument to its actual type. I think it is the most clean way to do all this stuff in one place. But of course I'm open for other solutions.

**Comment by islerfab** [comment]
I just wanted to write a proposal as a response to your comment, and saw that you've been very busy - thanks a lot! I like the new implementation much better, especially because you could get rid of the helper classes and thus fixed enumeration of AllocationLimit types that are possible. And actually, most of the "code smells" I was talking about were found in these helper classes, because that's exactly where such casts are necessary. I know that maps/lists and generics are a pain, but they often call it a feature of a strongly typed language, not a bug, since there's often slightly cleaner ways to solve it. Which you have found here.

As I understand it, you've resolved your concern about the good variables that may be missing from a WDP, right? A SATS domain would probably have to implement its own AllocationLimit class to validate, and take the allocation limits into account when creating the MIP, right?

Are you still worried about the performance? I don't see why the current implementation would be less performant than what you've had before - without having a deep understanding of it, it seems as if you've just shifted the responsibility from the various parts in the library towards a single module, but the constraints have to be created anyway... For the overall performance loss of creating these constraints and potentially making the MIP more complex, that's something we have to live with, the cost of having this feature...

> Just another small note. Random Sampling from a space limited with constraints on goods is far from trivial without enumerating all possible bundles. So before going for this solution we should think if this implementation is really what we want.

That is unfortunately true. A naïve way would be to resample if the current sample falls onto an invalid bundle. But of course that can be very inefficient in the edge case. That's probably where the question with the iterators would come up again... Do you think they should be part of the current effort? Or can you use a simple bundle sampling for your experiments for now, and then we'll implement an iterator logic as a next step? The obvious drawback of this would be that future experiments that use iterators wouldn't produce the same random bundles.

**Comment by beyeler** [comment]
> I just wanted to write a proposal as a response to your comment, and saw that you've been very busy - thanks a lot! I like the new implementation much better, especially because you could get rid of the helper classes and thus fixed enumeration of AllocationLimit types that are possible. And actually, most of the "code smells" I was talking about were found in these helper classes, because that's exactly where such casts are necessary. I know that maps/lists and generics are a pain, but they often call it a feature of a strongly typed language, not a bug, since there's often slightly cleaner ways to solve it. Which you have found here.

Yes and No. The new implementation delegates many features to the AllocationLimit class which do not strongly belong to it. We had this discussion about the implementation of Random Bundle sampling and where it belongs to. If you want the Random Sampling to happen in its own component and want to achive an efficient approach using an extensible component which holds the responsible Processors in Maps is a good approach. 

> As I understand it, you've resolved your concern about the good variables that may be missing from a WDP, right? A SATS domain would probably have to implement its own AllocationLimit class to validate, and take the allocation limits into account when creating the MIP, right?
> 
> Are you still worried about the performance? I don't see why the current implementation would be less performant than what you've had before - without having a deep understanding of it, it seems as if you've just shifted the responsibility from the various parts in the library towards a single module, but the constraints have to be created anyway... For the overall performance loss of creating these constraints and potentially making the MIP more complex, that's something we have to live with, the cost of having this feature...

The problem with the performance was never directly SATS or MLCA related as the respective MIPs contain good variables by definition. But in a general BidBased WinnerDetermination it is unefficient to encode goods, because they are not yet part of the MIP (they are encoded as constraints and not variables). While the first approach was able to add just another variable for the bundle (i.e. count the allocated number of items in this bundle). The new approach needs to encode each item contained in a bid/bundle separately. So in a domain like MRVM you might end up with 5000 x 96 additional variables in the ORWDP (10 bidders with 500 bids and up to 96 items for each bid).

> > Just another small note. Random Sampling from a space limited with constraints on goods is far from trivial without enumerating all possible bundles. So before going for this solution we should think if this implementation is really what we want.
> 
> That is unfortunately true. A naïve way would be to resample if the current sample falls onto an invalid bundle. But of course that can be very inefficient in the edge case. That's probably where the question with the iterators would come up again... Do you think they should be part of the current effort? Or can you use a simple bundle sampling for your experiments for now, and then we'll implement an iterator logic as a next step? The obvious drawback of this would be that future experiments that use iterators wouldn't produce the same random bundles.

If you have time it would be great to port the iterators to mechlib. However, it is not straight forward how an iterator with respect to an AllocationLimit should work (beside Uniform sampling, which is also the only thing I need for now).

File: src/main/java/org/marketdesignresearch/mechlib/core/bidder/Bidder.java (hunk 1 above)
Comment (beyeler) [reply src/main/java/org/marketdesignresearch/mechlib/core/bidder/Bidder.java:74]: done

**Comment by islerfab** [comment]
I agree that the allocation limit now has much, maybe "too" much responsibility. But I have a preference in this direction, because users that don't want allocation limits can just use the mechlib without ever having to care about the AllocationLimit's (growing) complexity.

For the inefficiency: I think there is an upper limit on the usability of the bid based OR-WDP in terms of #goods anyway. Do you think this is a realistic scenario, that MRVM is expressed in a non-generic sense (thus with 96 goods) and should be solved as an OR-WDP? If so, yes, we need to consider solving it in a different way.
When I was proposing this idea, btw, I had a less generic approach in mind than you've implemented now. I thought about an Interface AllocationLimit that defines methods for "adaptOrWDP(...)", "adaptXorWDP(...), etc., just all use cases an AllocationLimit has to cover inside the MechLib. This way, you could have the exact same efficient way to solve it as you've had before, if I'm not mistaken. I have not thought this through completely, but maybe you can figure out if that would be an option as well.

The iterators are not straight-forward in many ways... With or without AllocationLimits. I will have a look at it this week, probably Friday, if that's ok for you? I think adding yet another feature will delay your experiments, though, it's your call when you'd like to draw the line and release MechLib and SATS for your experiments.

**Comment by beyeler** [comment]
A user that does not use the AllocationLimit is affected by it the same way with with either implementation. The User only needs to take care of the AllocationLimit if he implements a new part that is may affected by it (e.g. a new WDP).

No there is no real limit for the ORWDP. It only contains 1 constraint for each good. But I agree that in practice CPLEX should be able to handle most domains of interest with AllocationLimits. 

I have a strong preference in separating functional implementations from logical constructs. Your concept is not really extensible without changing the mechlib itself, while my first approach was easily extensible with new concepts (e.g. WDPs) that depend on AllocationLimits (and also new AllocationLimits). Therefore the now proposed solution is the only reasable for me to implement this more generally.

The new implementation has some more draw backs (beside the RandomSampling we already discussed). The Allocation limit needs to know the complete set of goods (i.e. the Domain), but belongs to the bidder which itself belongs to a domain. Therefore I think it is no good style to ask in the constructor for the complete set of Goods available.

Concerning the iterators: I already started a first set of experiments, but as the cluster is busy at the moment I do not think I can start any expensive experiments before sunday. So therefore you are very welcome to have a look at it this week.

**Comment by islerfab** [comment]
As just discussed via call: Great work, I agree with you, and it seems like you've brought it to a point where you're also happy with it.

Regarding iterators, let's hold off that for now, since it will probably bring a chain of further required changes. The non-domain-specific iterators should be part of the MechLib at some point, and not of SATS, but this can wait for the version 0.1.1 of the mechlib.

File: src/main/java/org/marketdesignresearch/mechlib/core/allocationlimits/AllocationLimit.java [hunk 3]
```diff
@@ -0,0 +1,8 @@
+package org.marketdesignresearch.mechlib.core.allocationlimits;
+
+public interface AllocationLimit {
```
[the remark this answers is not in the document]
Comment (beyeler) [reply src/main/java/org/marketdesignresearch/mechlib/core/allocationlimits/AllocationLimit.java]: added TODO

File: src/main/java/org/marketdesignresearch/mechlib/core/allocationlimits/BundleSizeAllocationLimit.java [hunk 4]
```diff
@@ -0,0 +1,15 @@
+package org.marketdesignresearch.mechlib.core.allocationlimits;
+
+import org.marketdesignresearch.mechlib.core.bidder.Bidder;
+
+public interface BundleSizeAllocationLimit extends AllocationLimit{
```
[the remark this answers is not in the document]
Comment (beyeler) [reply src/main/java/org/marketdesignresearch/mechlib/core/allocationlimits/BundleSizeAllocationLimit.java]: added TODO

File: src/main/java/org/marketdesignresearch/mechlib/core/allocationlimits/GoodAllocationLimit.java [hunk 5]
```diff
@@ -0,0 +1,18 @@
+package org.marketdesignresearch.mechlib.core.allocationlimits;
+
+import java.util.List;
+
+import org.marketdesignresearch.mechlib.core.Good;
+
+public interface GoodAllocationLimit extends AllocationLimit{
```
[the remark this answers is not in the document]
Comment (beyeler) [reply src/main/java/org/marketdesignresearch/mechlib/core/allocationlimits/GoodAllocationLimit.java]: added TODO

File: src/main/java/org/marketdesignresearch/mechlib/core/allocationlimits/utils/AllocationLimitHelper.java [hunk 6]
```diff
@@ -0,0 +1,13 @@
+package org.marketdesignresearch.mechlib.core.allocationlimits.utils;
+
+import java.util.List;
+
+import org.marketdesignresearch.mechlib.core.Bundle;
+import org.marketdesignresearch.mechlib.core.Good;
+import org.marketdesignresearch.mechlib.core.allocationlimits.AllocationLimit;
+
+public interface AllocationLimitHelper<T extends AllocationLimit> {
```
[the remark this answers is not in the document]
Comment (beyeler) [reply src/main/java/org/marketdesignresearch/mechlib/core/allocationlimits/utils/AllocationLimitHelper.java]: added TODO

**Comment by beyeler** [comment]
Thx closing this branch without merge. All relevant changes are in Upstream branch Auction Random

**Pull request closed without merging**